designs/serialization.md

```diff
@@ -0,0 +1,296 @@
+according to some protocol, such as
+[AWS RestJson1](https://smithy.io/2.0/aws/protocols/aws-restjson1-protocol.html),
+based on information from a Smithy model.
+
+## Goals
+
+* Shared - Protocols should be implemented as part of a shared library. If two
+  clients using the same protocol are installed, they should use a shared
+  implementation. These implementations should be as compact as possible while
+  still being robust.
+* Hot-swappable - Implementations should be flexible enough to be swapped at
+  runtime if necessary. If a service supports more than one protocol, it should
+  be trivially easy to swap between them, even at runtime.
+* Flexible - Implementations should be useable for purposes other than as a
+  component of making a request to a web service. Customers should be able to
+  feed well-formed data from any source into a protocol and have it transform
+  that data with no side-effects.
+
+## Schemas
+
+The basic building block of Smithy is the "shape", a representation of data of a
+given type with known properties called "members", additional constraints and
+metadata called "traits", and an identifier.
+
+For each shape contained in a service, a `Schema` object will be generated that
+contains almost all of its information. Traits that are known to not affect
+serialization or deserialization will be omitted from the generated `Schema` to
+save space.
+
+Schemas will form the backbone of serialization and deserialization, carrying
+information that cannot be natively included in generated data classes.
+
+The `Schema` class will be a read-only dataclass. The following shows its basic
+definition, though the concrete definition may have a slightly different
+implementation and/or additional helper methods.
+
+```python
+@dataclass(kw_only=True, frozen=True)
+class Schema:
+    id: ShapeID
+    shape_type: ShapeType
+    traits: dict[ShapeID, "Trait"] = field(default_factory=dict)
+    members: dict[str, "Schema"] = field(default_factory=dict)
+    member_target: "Schema | None" = None
+    member_index: int | None = None
+
+    @classmethod
+    def collection(
+        cls,
+        *,
+        id: ShapeID,
+        shape_type: ShapeType = ShapeType.STRUCTURE,
+        traits: list["Trait"] | None = None,
+        members: Mapping[str, "MemberSchema"] | None = None,
+    ) -> Self:
+        ...
+
+
+@dataclass(kw_only=True, frozen=True)
+class Trait:
+    id: "ShapeID"
+    value: "DocumentValue" = field(default_factory=dict)
+```
+
+Below is an example Smithy `structure` shape, followed by the `Schema` it would
+generate.
+
+```smithy
+namespace com.example
+
+structure ExampleStructure {
+    member: Integer = 0
+}
+```
+
+```python
+EXAMPLE_STRUCTURE_SCHEMA = Schema.collection(
+    id=ShapeID("com.example#ExampleStructure"),
+    members={
+        "member": {
+            "target": INTEGER,
+            "index": 0,
+            "traits": [
+                Trait(id=ShapeID("smithy.api#default"), value=0),
```

Comment:
Document values don't need any kind of wrapping type?

Reply:
`DocumentValue` is the unwrapped json-like representation